api_docs/changelog.md

```diff
@@ -20,6 +20,11 @@ format used by the Zulip server that they are interacting with.
 
 ## Changes in Zulip 10.0
 
+**Feature level 298**
+
+* [`GET /users/{email}`](/api/get-user-by-email): TODO: write up details here and
+  zulip.yaml before merging.
```

Comment:
leaving the API docs out to not waste work if we end up tweaking the details of the behavior

Reply:
I think the API design is good and you can write this now accordingly.

Reply:
Cool, done. Feels a bit unclear what amount of detail is right to provide in which section of these docs since these details around `.email` are super fiddly (e.g. I don't want to flood the changelog with an essay comparing subtleties of how this works  now vs previously, but don't want to skip too much either). Made an attempt to balance it.

Reply:
I think it's pretty good; did some edits but it was quite quick to do so:
``` diff
diff --git a/api_docs/changelog.md b/api_docs/changelog.md
index e98b03e65e..c8fc2c6495 100644
--- a/api_docs/changelog.md
+++ b/api_docs/changelog.md
@@ -22,13 +22,9 @@ format used by the Zulip server that they are interacting with.
 
 **Feature level 302**
 
-* [`GET /users/{email}`](/api/get-user-by-email): Changed the `email` values
-  by which users can successfully be looked up to match the user email visibility
-  setting's semantics better. Now the real email of a user can be provided and the
-  endpoint will return the target user's data if and only if the target's email
-  visibility setting permits the requester to see the email address.
-  The dummy email addresses of the form `user{id}@{realm.host}` still work, and
-  will now work for **all** users, via identifying them by the embedded user ID.
+* [`GET /users/{email}`](/api/get-user-by-email): Changed the `email`
+  values by which users can successfully be looked up to match the
+  user email visibility setting's semantics better.
 
 **Feature level 301**
 
diff --git a/zerver/lib/users.py b/zerver/lib/users.py
index a289b7a9b4..06f20eb233 100644
--- a/zerver/lib/users.py
+++ b/zerver/lib/users.py
@@ -342,18 +342,25 @@ def access_user_by_email(
     allow_bots: bool = False,
     for_admin: bool,
 ) -> UserProfile:
-    """
-    Fetch a user by email address. Endpoints using this function can be queried either with:
-    1) The real email address of the intended user, if the requester believes that user exists and allows
-       their email address to be visible to the requester.
-    2) The dummy email address (of the approximate shape 'user<id>@<dummy email domain>') of the intended
-       user. We detect when the format of the provided email address matches the format of our dummy
-       email addresses and extract the user id from it for a regular id-based lookup.
-
-       In particular, this mode is kept around for backwards compatibility with the old behavior of the
-       `GET /users/{email}` endpoint, which required use of the dummy email address for lookups of any user
-       which didn't have EMAIL_ADDRESS_VISIBILITY_EVERYONE set, regardless of how the actual
-       email_address_visibility setting related to the role of the requester.
+    """Fetch a user by email address. Endpoints using this function can be queried either with:
+
+    1) The real email address of the intended user, if the requester
+       believes that user exists and allows their email address to be
+       visible to the requester via their `email_address_visibility` setting.
+
+    2) The dummy email address (of the approximate shape
+       'user{user_id}@{realm_dummy_email_domain}') of the intended user. We
+       detect when the format of the provided email address matches
+       the format of our dummy email addresses and extract the user id
+       from it for a regular id-based lookup.
+
+       In particular, this mode is kept around for backwards
+       compatibility with the old behavior of the `GET /users/{email}`
+       endpoint, which required use of the dummy email address for
+       lookups of any user which didn't have
+       EMAIL_ADDRESS_VISIBILITY_EVERYONE set, regardless of how the
+       actual email_address_visibility setting related to the role of
+       the requester.
 
        Note: If the realm.host value changes (e.g. due to the server moving to a new
        domain), the required dummy email values passed here will need to be updated
@@ -382,15 +389,25 @@ def access_user_by_email(
             for_admin=for_admin,
         )
 
-    # We're dealing with a real email address, so we need to do a careful lookup,
-    # ensuring to abide by email address visibility restrictions.
+    # Since the format doesn't match, we should treat it as a lookup
+    # for a real email address.
     allowed_email_address_visibility_values = (
         UserProfile.ROLE_TO_ACCESSIBLE_EMAIL_ADDRESS_VISIBILITY_IDS[user_profile.role]
     )
 
     try:
-        # Fetch the user from the subset of users which allow the requester to see their email address.
-        target = UserProfile.objects.get(
+        # Fetch the user from the subset of users which allow the
+        # requester to see their email address. We carefully do this
+        # with a single query to hopefully make timing attacks
+        # ineffective.
+        #
+        # Notably, we use the same select_related as access_user_by_id.
+        target = UserProfile.objects.select_related(
+            "realm",
+            "realm__can_access_all_users_group",
+            "realm__can_access_all_users_group__named_user_group",
+            "bot_owner",
+        ).get(
             delivery_email__iexact=email.strip(),
             realm=user_profile.realm,
             email_address_visibility__in=allowed_email_address_visibility_values,
diff --git a/zerver/openapi/zulip.yaml b/zerver/openapi/zulip.yaml
index 1443d871dc..9f6f7c4f7d 100644
--- a/zerver/openapi/zulip.yaml
+++ b/zerver/openapi/zulip.yaml
@@ -12398,17 +12398,22 @@ paths:
           description: |
             The email address of the user to fetch. Two forms are supported:
 
-            - The real email address of the user. The lookup will succeed if and only if
-              the user exists and their email address visibility setting permits the client
-              to see the email address.
-            - The dummy email address of the form `user{id}@{realm.host}`. This is identical
-              to simply [getting user by ID](/api/get-user). If the server or realm change
-              domains, the dummy email address used has to be adjustment to match the new
-              realm domain.
+            - The real email address of the user (`delivery_email`). The lookup will
+              succeed if and only if the user exists and their email address visibility
+              setting permits the client to see the email address.
+
+            - The dummy Zulip API email address of the form `user{user_id}@{realm_host}`. This
+              is identical to simply [getting user by ID](/api/get-user). If the server or
+              realm change domains, the dummy email address used has to be adjustment to
+              match the new realm domain. This is legacy behavior for
+              backwards-compatibility, and will be removed in a future release.
 
             **Changes**: Starting with Zulip 10.0 (feature level 302), lookups by real email
             address match the semantics of the target's email visibility setting and dummy
-            email addresses work for all users, independently of their email visibility setting.
+            email addresses work for all users, independently of their email visibility
+            setting.
+
+            Previously, lookups were done only using the Zulip API email addresses.
           schema:
             type: string
           example: iago@zulip.com
```